src/map/entities/petentity.cpp

```diff
@@ -206,12 +237,58 @@ void CPetEntity::Spawn()
         mobutils::GetAvailableSpells(this);
     }
 
+    if (m_PetType == PET_TYPE::JUG_PET)
+    {
+        m_jugSpawnTime = server_clock::now();
+    }
+
     // NOTE: This is purposefully calling CBattleEntity's impl.
     // TODO: Calling a grand-parent's impl. of an overridden function is bad
     CBattleEntity::Spawn();
     luautils::OnMobSpawn(this);
 }
 
+bool CPetEntity::shouldPersistThroughZone()
+{
+    // TODO: Add other types of pets that should persist through zone here
+    return getPetType() == PET_TYPE::WYVERN ||
+           getPetType() == PET_TYPE::AUTOMATON ||
+           (getPetType() == PET_TYPE::JUG_PET && settings::get<bool>("map.KEEP_JUGPET_THROUGH_ZONING"));
+}
+
+bool CPetEntity::shouldDespawn(time_point tick)
+{
+    if (isCharmed && tick > charmTime)
+    {
+        return true;
+    }
+
+    if (PMaster != nullptr &&
+        PAI->IsSpawned() &&
```

Comment:
Note that PAI->IsSpawned() could be moved to the top of this function, but I'm not sure of the logic of charmed pets and did not want to change it.